internal/cmd/csvgen/main.go

```diff
@@ -0,0 +1,463 @@
+				Series: []flux.GroupKey{k},
+			})
+		}
+		return groups
+	}
+
+	var groups []SeriesGroup
+	mapping := make(map[string]*SeriesGroup)
+	for _, k := range keys {
+		parts := make([]string, 0, len(by))
+		for _, s := range by {
+			idx := execute.ColIdx(s, k.Cols())
+			if idx == -1 {
+				continue
+			}
+			parts = append(parts, k.ValueString(idx))
+		}
+
+		if len(parts) == 0 {
+			groups = append(groups, SeriesGroup{
+				Series: []flux.GroupKey{k},
+			})
+			continue
+		}
+
+		groupKey := strings.Join(parts, ",")
+		gr, ok := mapping[groupKey]
+		if !ok {
+			groups = append(groups, SeriesGroup{
+				Series: []flux.GroupKey{k},
+			})
+			mapping[groupKey] = &groups[len(groups)-1]
+			continue
+		}
+		gr.Series = append(gr.Series, k)
+	}
+	return groups
+}
+
+type DataGenerator struct {
+	Start     values.Time
+	Period    values.Duration
+	Jitter    values.Duration
+	NumPoints int
+}
+
+func (dg *DataGenerator) Generate(tb execute.TableBuilder, r *rand.Rand, typ flux.ColType) (start, stop values.Time) {
+	var next func() values.Value
+	switch typ {
+	case flux.TFloat:
+		next = func() values.Value {
+			v := rand.NormFloat64() * 50
+			return values.NewFloat(v)
+		}
+	case flux.TInt:
+		next = func() values.Value {
+			v := rand.Intn(201) - 100
+			return values.NewInt(int64(v))
+		}
+	case flux.TUInt:
+		next = func() values.Value {
+			v := rand.Intn(101)
+			return values.NewUInt(uint64(v))
+		}
+	case flux.TString:
+		next = func() values.Value {
+			v := genTagValue(r, 3, 8)
+			return values.NewString(v)
+		}
+	case flux.TBool:
+		next = func() values.Value {
+			v := r.Intn(2) == 1
+			return values.NewBool(v)
+		}
+	default:
+		panic("implement me")
+	}
+
+	timeIdx, _ := tb.AddCol(flux.ColMeta{
+		Label: execute.DefaultTimeColLabel,
+		Type:  flux.TTime,
+	})
+	valueIdx, _ := tb.AddCol(flux.ColMeta{
+		Label: execute.DefaultValueColLabel,
+		Type:  typ,
+	})
+
+	start, stop = dg.Start, dg.Start
+	for i := 0; i < dg.NumPoints; i++ {
+		ts := dg.Start.Add(values.Duration(i) * dg.Period)
+		if dg.Jitter != 0 {
+			jitter := r.Intn(int(dg.Jitter)*2 + 1)
+			ts = ts.Add(values.Duration(jitter))
+		}
+		_ = tb.AppendTime(timeIdx, ts)
+		_ = tb.AppendValue(valueIdx, next())
+		if ts > stop {
+			stop = ts
+		}
+	}
+	return start, stop
+}
+
+func getTags() (map[string]int, error) {
+	if len(Flags.Keys) == 0 {
+		return nil, nil
+	}
+
+	m := make(map[string]int)
+	keys := strings.Split(Flags.Keys, ",")
+
+	if len(Flags.TagCardinality) == 0 {
+		for _, k := range keys {
+			m[k] = 1
+		}
+		return m, nil
+	}
+
+	cardinality := strings.Split(Flags.TagCardinality, ",")
+	if len(cardinality) != len(keys) {
+		return nil, fmt.Errorf("cardinality must have the same number of entries as keys")
+	}
+	for i, k := range keys {
+		c, err := strconv.Atoi(cardinality[i])
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse %s as an integer: %s", cardinality[i], err)
+		}
+		m[k] = c
+	}
+	return m, nil
+}
+
+func main() {
```

Comment:
I'd really like to see this implemented as a source type inside of `flux` instead.  reasons in my full review, but most of the code here could go into a `Decode` method as shown here:  https://github.com/influxdata/flux/blob/c45e00c1dd69f8143edf0daff5e1a5efa98a446f/functions/inputs/from_generator.go#L164